Homeworks/homework_14/test_site.py

```diff
@@ -0,0 +1,52 @@
+import pytest
+from selenium import webdriver
+from selenium.common import NoSuchElementException
+import fixture
+from fixture import driver_init, login_data, driver_init, setup
+
+
+@pytest.mark.usefixtures("setup")
+class TestLogin:
+
+    @pytest.fixture(scope="function", autouse=True)
+    def test_login_page_displayed(self):
+        self.login_page.open_tab_page()
+        assert self.login_page.is_login_page_username_displayed()
+        assert self.login_page.is_login_page_password_displayed()
+
+    def test_successful_login(self, request, login_data):
+        self.login_page.login(user=login_data["user"], password=login_data["password"])
+        self.login_page.click_on_login_btn()
+        assert "Log out" in self.login_page.get_logout_button_text().strip()
+        assert "Welcome a_test" in self.login_page.get_welcome_text().strip()
+
+
+@pytest.mark.usefixtures("setup")
+class TestCategory:
+
+    def pre_req(self, login_data):
+        self.login_page.open_tab_page()
+        self.login_page.login(user=login_data["user"], password=login_data["password"])
+        self.login_page.click_on_login_btn()
+
+    def test_product_name_and_price(self, login_data):
+        self.pre_req(login_data)
```

Comment:
It would be better to use fixture to implement test precondition (or teardown) instead of using separate test function pre_req. The second approach causes inconsistency because test function is expected to be a representation of a specific test scenario and not other entities.

Reply:
pre_req fixture was created for TestCategory 
Thank you!